### PR TITLE
Add Decimal support to set_agg and set_union

### PR DESCRIPTION
**Pull request by rrando901**
Delivers https://github.com/facebookincubator/velox/issues/7935

### CODE REVIEW

File: velox/dwio/dwrf/writer/ColumnWriter.cpp
Comment (aditi-pandit) [inline velox/dwio/dwrf/writer/ColumnWriter.cpp:1]: Please can you separate the changes for ColumnWriter in another PR and add tests for it as well.

File: velox/functions/prestosql/aggregates/SetAggregates.cpp [hunk 1]
```diff
@@ -398,6 +398,11 @@ std::unique_ptr<exec::Aggregate> create(
       return std::make_unique<Aggregate<int32_t>>(resultType);
     case TypeKind::BIGINT:
       return std::make_unique<Aggregate<int64_t>>(resultType);
+    case TypeKind::HUGEINT:
+      if(inputType->isLongDecimal()) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/aggregates/SetAggregates.cpp:402]: You can add VELOX_CHECK(inputType->isLongDecimal()) before the if statement instead of the VELOX_NYI.

File: velox/dwio/dwrf/writer/ColumnWriter.cpp [hunk 2]
```diff
@@ -2031,6 +2031,12 @@ std::unique_ptr<BaseColumnWriter> BaseColumnWriter::create(
     case TypeKind::BIGINT:
       return std::make_unique<IntegerColumnWriter<int64_t>>(
           context, type, sequence, onRecordPosition);
+    case TypeKind::HUGEINT:
+      if(type.type()->isLongDecimal()) {
```
Comment (aditi-pandit) [inline velox/dwio/dwrf/writer/ColumnWriter.cpp:2035]: You can add VELOX_CHECK(type.type()->isLongDecimal()) instead of raising an exception here.

File: velox/functions/prestosql/aggregates/SetAggregates.cpp [hunk 3]
```diff
@@ -458,6 +464,11 @@ void registerSetAggAggregate(const std::string& prefix) {
             return std::make_unique<SetAggAggregate<int32_t>>(resultType);
           case TypeKind::BIGINT:
             return std::make_unique<SetAggAggregate<int64_t>>(resultType);
+          case TypeKind::HUGEINT:
+            if(inputType->isLongDecimal()) {
+              return std::make_unique<SetAggAggregate<int128_t>>(resultType);
+            }
+            VELOX_NYI("Non-decimal use of HUGEINT is not supported");
```
Comment (aditi-pandit) [inline velox/functions/prestosql/aggregates/SetAggregates.cpp:471]: Same here. Prefer the CHECK over NYI.

File: velox/functions/prestosql/aggregates/tests/SetAggTest.cpp [hunk 4]
```diff
@@ -155,6 +155,90 @@ TEST_F(SetAggTest, groupBy) {
       {expected});
 }
 
+TEST_F(SetAggTest, decimal) {
+  // Test with short decimal
+  auto type = DECIMAL(6,2);
+
+  auto data = makeRowVector({
+      makeFlatVector<int64_t>({1000, 2000, 3000, 4000, 5000, 3000, 4000, 2000, 6000, 7000}, type),
```
Comment (aditi-pandit) [inline velox/functions/prestosql/aggregates/tests/SetAggTest.cpp:163]: Please add values at the boundaries on int64_t. Also add -ve numbers in the tests as well.

**Comment by yzhang1991** [comment]
@aditi-pandit Can you help take a look?

File: velox/functions/prestosql/aggregates/tests/SetAggTest.cpp [hunk 5]
```diff
@@ -155,6 +160,185 @@ TEST_F(SetAggTest, groupBy) {
       {expected});
 }
 
+TEST_F(SetAggTest, decimal) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/aggregates/tests/SetAggTest.cpp:163]: This method has become too long. Split this test into 2 for shortDecimal and longDecimal.

**Review by aditi-pandit — COMMENTED** [review]
Thanks @rrando901. One minor comment.

File: velox/functions/prestosql/aggregates/tests/SetAggTest.cpp (hunk 5 above)
Comment (rrando901) [reply velox/functions/prestosql/aggregates/tests/SetAggTest.cpp]: Split the test into 2 for shortDecimal and longDecimal.

File: velox/functions/prestosql/aggregates/tests/SetUnionTest.cpp [hunk 6]
```diff
@@ -324,5 +329,175 @@ TEST_F(SetUnionTest, inputOrder) {
   assertQuery(plan, expected);
 }
 
+TEST_F(SetUnionTest, decimal) {
```
[the remark this answers is not in the document]
Comment (rrando901) [reply velox/functions/prestosql/aggregates/tests/SetUnionTest.cpp]: Split the test into 2 for shortDecimal and longDecimal.

**Review by aditi-pandit — APPROVED** [review]
Thanks @rrando901

**Comment by aditi-pandit** [comment]
@rrando901 : Your build has a fuzzer failure. Please can you rebase your code and try again. Also create an issue for the fuzzer failure if needed.

**Comment by rrando901** [comment]
@aditi-pandit Rebased code with the latest from main. There is a fuzzer failure in `AsyncDataCacheTest.cpp`.

Looks like there are two issues opened for this already:
- https://github.com/facebookincubator/velox/issues/8189
- https://github.com/facebookincubator/velox/issues/3968

**Comment by rrando901** [comment]
Hi @Yuhta. When you get a chance, could you take a look at these changes?

**Comment by rrando901** [comment]
Rebased PR @aditi-pandit @Yuhta
